integration/fuse/src/main/java/alluxio/fuse/file/FuseFileStream.java

```diff
@@ -0,0 +1,134 @@
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.fuse.file;
+
+import static jnr.constants.platform.OpenFlags.O_ACCMODE;
+
+import alluxio.AlluxioURI;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.fuse.AlluxioFuseUtils;
+import alluxio.fuse.auth.AuthPolicy;
+
+import jnr.constants.platform.OpenFlags;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This interface should be implemented by all fuse file streams.
+ */
+public interface FuseFileStream extends AutoCloseable {
+
+  /**
+   * Reads data from the stream.
+   *
+   * @param buf the byte buffer to read data to
+   * @param size the size to read
+   * @param offset the offset to read
+   * @return the bytes read
+   */
+  int read(ByteBuffer buf, long size, long offset);
+
+  /**
+   * Writes data to the stream.
+   *
+   * @param buf the byte buffer to read data from and write to the stream
+   * @param size the size to write
+   * @param offset the offset to write
+   */
+  void write(ByteBuffer buf, long size, long offset);
+
+  /**
+   * Gets the file length.
+   *
+   * @return file length
+   */
+  long getFileLength();
+
+  /**
+   * Flushes the stream.
+   */
+  void flush();
+
+  /**
+   * Truncates the file to the given size.
+   *
+   * @param size the truncate size
+   */
+  void truncate(long size);
+
+  /**
+   * Closes the stream.
+   */
+  void close();
+
+  /**
+   * Factory for {@link FuseFileInStream}.
+   */
+  @ThreadSafe
+  class Factory {
+    private final FileSystem mFileSystem;
+    private final AuthPolicy mAuthPolicy;
+
+    /**
+     * Creates an instance of {@link FuseFileStream.Factory} for
+     * creating fuse streams.
+     *
+     * @param fileSystem the file system
+     * @param authPolicy the authentication policy
+     */
+    public Factory(FileSystem fileSystem, AuthPolicy authPolicy) {
+      mFileSystem = fileSystem;
+      mAuthPolicy = authPolicy;
+    }
+
+    /**
+     * Factory method for creating an implementation of {@link FuseFileStream}.
+     *
+     * @param uri the Alluxio URI
+     * @param flags the create/open flags
+     * @param mode the create file mode, -1 if not set
+     * @param status the uri status
+     * @return the created fuse file stream
+     */
+    public FuseFileStream create(
+        AlluxioURI uri, int flags, long mode, Optional<URIStatus> status) {
+      switch (OpenFlags.valueOf(flags & O_ACCMODE.intValue())) {
+        case O_RDONLY:
+          return FuseFileInStream.create(mFileSystem, uri, flags, status);
+        case O_WRONLY:
+          return FuseFileOutStream.create(mFileSystem, mAuthPolicy, uri, flags, mode, status);
+        case O_RDWR:
```

Comment:
I see this flag, this means it can still return a stream, but when the user tries to read, it will throw an exception,  right?

Reply:
This is the FuseFileInOrOutStream, 
when user do the first read() operation, it's read
if user do the first write() related operation e.g. truncate() or write(), it's write

Reply:
Got it.